File: heap_PriorityQueue/cppRealizations/lastStoneWeight.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;


class Solution {
public:
    int lastStoneWeight(vector<int>& stones) {
        priority_queue<int> pq(stones.begin(), stones.end());
        while(pq.size() > 1){
            int a = pq.top();
            pq.pop();
            int b = pq.top();
            pq.pop();
            if(a > b){
                pq.push(a - b);
            }
        }
        if(pq.size() == 0){
            return 0;
        }
        return abs(pq.top());
    }
};


int main() {
    vector<int> stones = {2, 7, 4, 1, 8, 1};
    int result = Solution().lastStoneWeight(stones);
    cout << result << endl;
    return 0;
}

File: heap_PriorityQueue/lastStoneWeight.py
import heapq


class Solution:
    def lastStoneWeight(self, stones: list[int]) -> int:
       stones = [-x for x in stones]
       heapq.heapify(stones)
       while len(stones) > 1:
           first = -heapq.heappop(stones)
           second = -heapq.heappop(stones)
           if first > second:
               heapq.heappush(stones, second - first)
       stones.append(0)
       return abs(stones[0])


if __name__ == "__main__":
    stones = [2,7,4,1,8,1]
    print(Solution().lastStoneWeight(stones))